active/0004-integrating-extensible-resources-and-deployment-stacks.md

```diff
@@ -0,0 +1,547 @@
+
+The flow is a non-standard OAuth 2.0 extension developed by Microsoft.
+
+## Motivation
+
+Deployment Stacks offer users the convenience of managing a set of resources as a unified entity, enabling quicker updates and deletions. Currently, this functionality is limited to Azure resources due to the absence of a unique identifier (resource ID) for extensible resources and a standardized authentication mechanism across different extensibility providers.
+
+To overcome these limitations, the REP suggests modifications to the deployments API. A proposed change is the introduction of a `referenceID` field in `outputResources`, a freefrom string designed to uniquely identify extensible resources. Additionally, the proposal includes updates to the `deploymentProviders`, enabling it to store provider configurations. These configurations may include references to key vaults or resource properties, which can be utilized to access authentication details safely. These proposed enhancements allows Deployment Stacks to interact with extensibility providers to delete extensible resources.
+
+Implementing these changes ensure that extensible resources can be managed with the same level of efficiency and security as Azure resources within Deployment Stacks.
+
+## Detailed design
+
+### Reference ID
+
+> A reference ID is a URL-encoded, free-form string designed to uniquely identify an extensible resource. This ID is generated by extensibility providers and is returned as part of the responses from the `POST /resources/reference` or `POST /resources/preview` extensible resource APIs. The concept and design of reference IDs have been finalized.
+>
+> Any reference ID MUST adhere to the format `{flags}-{uniqueId}`.
+
+> - `{flags}`: This section consists of two binary digits indicating the type of reference ID. The currently accepted values are `00` and `01`:
+>   - `00`: signifies that the reference ID is autonomous and doesn't need extra data for GET and DELETE actions. This implies the utilization of the extensible resource APIs `GET /resources/{referenceId}` and `DELETE /resources/{referenceId}` for operations.
+>   - `01`: indicates that additional information, specifically user-provided provider configuration (like a connection string), is essential alongside the reference ID to execute GET and DELETE operations. In this case, APIs `POST /resources/{referenceId}/getWithConfig` and `POST /resources/{referenceId}/deleteWithConfig` must be used, and the provider configuration must be included in the request body.
+> - `{uniqueId}`: This is a URL-encoded free-form string that serves to uniquely identify a resource. The format of uniqueId is defined by the providers and is opaque to the Azure Deployments engine.
+>
```

Comment:
It looks like the reference ID is passed in the URL - not the unique ID. If my unique ID for a resource is "a-cool-resource" then the requests would look like:

- `POST /resources/00-a-cool-resource`
- `DELETE /resources/01-a-cool-resource/deleteWithConfig`

It feels like the "flags" in this case are really serving as an API-version.